Live suggestions for sample type input

gregdingle opened the pull request:
# Description

This changes the sample type input to be a live autocomplete similar to the location input. The data comes from newly added `SampleType`. Unlike `location`, all possible suggestions are returned from the server and the filtering is done in JS alone, because of the small number of possible suggestions.

Also, there are two categories of match: "suggested" and "all". Suggested depends on the host genome, whether it is a human or insect or neither. This follows the framework created by @katrinakalantar at https://docs.google.com/spreadsheets/d/1_hPkQe5LI0Zw_C0Ls4HVCEDsc_FNNVOaU_aAfoaiZRE/edit?ts=5e152439#gid=0 . 

Plain text entries are still possible.

# Notes

* Although the suggested values come from a rails model, the selected values are stored as more generic `Metadatum`. 
* "OTHER" would be a more accurate description, but "ALL" is the word in the design at https://chanzuckerberg.invisionapp.com/share/A6V572ZSBU5#/screens/396544828 .
* I came up with the string matching algorithm based on intuition
* `MetadataInput` component is growing somewhat out of control because all metadata type specific props are being passed in: `isHuman`, `sampleTypes`, etc
* I added to the host genome wiki: https://github.com/chanzuckerberg/idseq-web/wiki/%5BDev%5D-How-to-update-host-genomes#web-app-update. "- (Uncommon) If your host is an insect, add it to the constant `INSECT_GENOME_IDS` so that only insect-appropriate sample types will be suggested for it."
* Many of the official types are too large for the input box so they are clipped. The input is the same width as before so I left it. 

# Tests

Try "apply all" 
Try saving and reloading

Try a variety of inputs. See expected results in suggested, all and plain text.

Check full word match, lowercase
![image](https://user-images.githubusercontent.com/28797/72028888-0bc1db00-3239-11ea-9212-770c9515c35c.png)

Check insect body parts for insect host
![image](https://user-images.githubusercontent.com/28797/72028891-0d8b9e80-3239-11ea-9351-9cb132975a0f.png)

Check "CSF" acronym match
![image](https://user-images.githubusercontent.com/28797/72028895-0f556200-3239-11ea-83e9-11c8d304bcf8.png)

Check second word in two words
![image](https://user-images.githubusercontent.com/28797/72028897-10868f00-3239-11ea-9f12-194ce815cf79.png)

Check prefix
![image](https://user-images.githubusercontent.com/28797/72028900-12e8e900-3239-11ea-8c44-0406b8f303ad.png)

gregdingle commented on app/assets/src/components/ui/controls/SampleTypeSearchBox.jsx:
```diff
@@ -0,0 +1,93 @@
+    }
+    // Always lowercase plain text input to distinguish it from Title Case
+    // offical sample types, which are enforced in SampleType.rb.
+    // TODO (gdingle): enforce lowercasing in CSV upload as well.
```
not sure how best to do the same for CSV and CLI uploads in a simple way, and not sure if it's worth the trouble... @MarkAZhang ?

tfrcarvalho commented on app/assets/src/components/common/MetadataManualInput.jsx:
```diff
@@ -29,6 +29,13 @@ import MetadataInput from "./MetadataInput";
 
 class MetadataManualInput extends React.Component {
+  // List as of 2020-01-08.
+  INSECT_GENOME_IDS = [
```
Did you consider including this info in the DB schema somehow (host genome model)? We might start missing updating this when we add new insect genomes.

gregdingle replied:
Yes, I considered it... and I think that it is reasonable to add a column such as `sample_type_taxa_name` to `host_genomes`. However, I decided against it for this PR because:
* we haven't had a chance to play with the suggestions yet and there may be bigger changes that result 
* the hard-coded precedent in the code of `isHuman` 
* the cost of a missed insect is low: the same options would be available in the "ALL" category instead of "SUGGESTED"

tfrcarvalho commented on app/assets/src/components/ui/controls/SampleTypeSearchBox.jsx:
```diff
@@ -0,0 +1,93 @@
+    const isSuggested = sampleType => {
+      // Insects only get their own types
+      if (this.props.isInsect) {
+        return sampleType.insect_only ? "suggested" : "donotshow";
```
These string should probably be constants.

gregdingle replied:
No problem. Side note: it's funny how we are more strict about constants in javascript than in python.

tfrcarvalho replied:
I am guessing you meant ruby... when i notice the pattern i try to push for it on both. Not sure if I have been consistent though :P

gregdingle replied:
No, I meant python, in the pipeline. It's full of magic strings. I think in ruby we consistently use constants.

tfrcarvalho commented on app/assets/src/components/ui/controls/SampleTypeSearchBox.jsx:
```diff
@@ -0,0 +1,93 @@
+  handleSearchTriggered = query => {
+    const matchSampleTypes = sampleType => {
+      // Match chars in any position. Good for acronyms.
+      const regex = new RegExp(query.split("").join(".*"), "gi");
```
are we assuming single word? What happens with spaces and word order?

gregdingle replied:
Currently, spaces are considered just another character, so the query cannot contain more spaces than in the name. For example: "ab domen" will not match. "whole b" will match. 

Having thought about it now, I think ignoring spaces in the query is more intuitive. What do you think?

tfrcarvalho replied:
Sorry, did not quite get the example: what does "whole b" match?

tfrcarvalho replied:
I see, "whole blood". 
I guess its more of a UX decision. IMO, I would probably split it on spaces and run the query independently for each word. Would love that "blood whole" would match too... (maybe there are other side effects im not seeing right now).

gregdingle replied:
I changed it to ignore the spaces. As for matching words in any order, it seems like overkill to me, but happy to be convinced otherwise.

tfrcarvalho replied:
For this particular case, it probably is... for a more general function I would do it. So, up to you.

MarkAZhang commented:
`<MetadataInput>` is also used in the `MetadataTab` which is in the sample details sidebar. You should also verify that your new input works there as expected.

MarkAZhang commented on app/assets/src/components/ui/controls/SampleTypeSearchBox.jsx:
```diff
@@ -0,0 +1,94 @@
+      }
+      // Humans get any type
+      if (this.props.isHuman) {
+        return this.SUGGESTED;
```
Nit: Could we move the constants above the class and just have `SUGGESTED` instead of `this.SUGGESTED`?

gregdingle replied:
Good point. I guess that is more idiomatic.

gregdingle commented on app/assets/src/components/ui/controls/SampleTypeSearchBox.jsx:
```diff
@@ -0,0 +1,93 @@
+    };
+    const matchedSampleTypes = this.props.sampleTypes.filter(matchSampleTypes);
+
+    const isSuggested = sampleType => {
```
Good point. `isSuggested` was leftover from when the function was returning a boolean!

gregdingle commented on app/assets/src/components/utils/propTypes.js:
```diff
@@ -141,5 +148,6 @@ export default {
   SampleUploadType,
   DownloadType,
   DownloadTypeParam,
+  SampleTypeType,
```
I think `SampleType` would be a mistake because of the ambiguity with `Sample`. `SampleTypeType` just seems funny to me, but how about `SampleTypeProps`?

gregdingle commented on app/assets/src/components/common/MetadataManualInput.jsx:
```diff
@@ -29,6 +29,13 @@ import MetadataInput from "./MetadataInput";
 
 class MetadataManualInput extends React.Component {
+  // List as of 2020-01-08.
+  INSECT_GENOME_IDS = [
```
The only reasons are what I put above. FYI, I followed the precedent of `isHuman={sampleHostGenomeId === 1}`

gregdingle commented on db/migrate/20200108180542_create_sample_types.rb:
```diff
@@ -27,6 +27,7 @@ def populate_sample_types!
       'CNS Infections': [
         { name: 'Brain' },
         { name: 'Cerebrospinal Fluid', human_only: true },
+        { name: 'Central Nervous System' },
```
I opportunistically fixed this following discussion with Katrina . It will work as long as this PR is deployed along with the previous 😸

gregdingle commented:
> MetadataInput is also used in the MetadataTab which is in the sample details sidebar. You should also verify that your new input works there as expected.

Good point. I'll check that now. 

> If the user clicks out of the input without picking something, the text stays in the input but nothing is actually selected (which isn't indicated). Might want to select-on-blur like the location box.

Good catch! This was actually a subtle bug. 

> Let's say you have both insect and human samples in your upload. You pick a human-only sample type for a human sample and click "Apply to All". This would apply the human-only sample type to all the insect samples as well. Not sure if this is an issue.

This would not be much of an issue in our current design, because the pairing of host and sample types is a suggestion only.

katrinakalantar commented on db/migrate/20200108180542_create_sample_types.rb:
```diff
@@ -27,6 +27,7 @@ def populate_sample_types!
+        { name: 'Central Nervous System' },
         { name: 'Ocular Fluid', human_only: true },
       ],
 
```
May make mores sense (for consistency across the sample groups) to use 
- "Central Nervous System" instead of "CNS Infections"
- "Respiratory Tract" instead of "Respiratory Infections" 
- "Reproductive Tract" instead of "Reproductive"

gregdingle commented:
@tfrcarvalho @MarkAZhang Updates: 

* I fixed the issues that @MarkAZhang raised
* I tested the input on the single sample page
![image](https://user-images.githubusercontent.com/28797/72112920-d6c68e80-32f3-11ea-92ba-3f28c7cbda1e.png)
* I went ahead and added a new column `taxa_category` to `host_genomes` as discussed, hooked up the UI, and I replaced pre-existing `isHuman` logic, so now we can support multiple "human" genomes if we want.

MarkAZhang commented on app/models/host_genome.rb:
```diff
@@ -2,6 +2,13 @@ class HostGenome < ApplicationRecord
+    # https://docs.google.com/spreadsheets/d/1_hPkQe5LI0Zw_C0Ls4HVCEDsc_FNNVOaU_aAfoaiZRE/
+    "human",
+    "insect",
+  ], }, if: -> { respond_to?(:taxa_category) } # for migrations
```
Wow! Cool solution.
